### DISCUSSION TITLE
Eversense CGM integration via NotificationListenerService

### DISCUSSION THREAD

**Discussion by BernhardRo** [NightscoutFoundation/xDrip#3290] [Ideas]
Hello, would it be possible to extend your UiBasedCollector to work with Eversense as well? Eversense also implements a permanent notification, which can be identified by "com.senseonics.gen12androidapp". The current glucose value is then available as string in "notification.tickerText"
Right now, a patch for the Eversense App exists to access the data, but with every update of the App, it gets more and more critical to patch it!

**Comment by Navid200** [comment]
Can I ask a question for my benefit to see if I can personally help or not?

The UI based collector, by default, does no backfill.  So, if you walk away from your phone, when you come back, you will have missing readings in xDrip.
There is an exception for some particular devices like Dexcom transmitters.  But, for others, that's the limitation.

Can you tell me if the existing Eversense hardware database in xDrip offers backfill or not?

My second question is about the app notification you have mentioned.
Can you post a screenshot of the notification?
What is the name of the app in the play store so that I can install for testing?
Does the app produce a unique notification for each reading showing the reading?

**Reply by BernhardRo** [reply]
Wow, that was a fast reaction!

The solution we have right now supports backfilling. But as I mentioned, I am not sure if we will be able to patch future versions as well. So to have an alternative, easy way of access (which works out of the box with the original app) would be a huge advantage.
The app can be found here: https://play.google.com/store/apps/details?id=com.senseonics.gen12androidapp
There is also a specific US version, but that not accessible from Europe (but I would have the apk)
I have tested it with https://github.com/pilot51/voicenotify (modified) and I got readings probably every 10sec - even though the value itself is only updated every 5min. 
This is how it looks like:
![Screenshot (12 01 2024 21_32_20)](https://github.com/NightscoutFoundation/xDrip/assets/35614076/c78dbd55-62d3-4f25-bb4e-4c2a759e9a92)

**Reply by Navid200** [reply]
Will you be able to do a test for us?
If you install the official app that you have access to, and set xDrip to the companion app hardware data source, does it not get readings?

**Reply by BernhardRo** [reply]
Up to now, I do not get any values.

**Reply by Navid200** [reply]
Can you post a screenshot of the classic status page?
https://navid200.github.io/xDrip/docs/xDrip-Version.html

**Comment by BernhardRo** [comment]
I have downloaded and installed xDrip from here: https://jamorham.github.io/#xdrip-plus (~2h ago) and not sure, if this is the correct version.
At least, I do not get values.
I updated the permissions as it requires access to the notifications. This was not directly possible on my Pixel 6a with Android 14. Instead, I had to allow restricted settings before, see https://support.google.com/android/answer/12623953?hl=en
I am not able to compile xDrip on my machine, otherwise I would help with it. But when I take a look at this voicenotify with the Debugger, it looks like this: 
<img width="401" alt="image" src="https://github.com/NightscoutFoundation/xDrip/assets/35614076/b20a62f5-84ec-4eff-b2bd-bef5690c6d2f">

**Reply by Navid200** [reply]
Can you try the latest Nightly?
This is the latest nightly: https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.01.13

I have gone over the entire update/download/install procedure and options here: 
https://navid200.github.io/xDrip/docs/Updates.html

You just need to download the file with apk suffix under assets and install over your existing version.

Does it then work with companion app mode and official Eversense app?

**Reply by BernhardRo** [reply]
I have updated it to the last nightly, but still do not get values:

![Screenshot_20240113-181008](https://github.com/NightscoutFoundation/xDrip/assets/35614076/29994870-d888-42da-8f6d-f8e8d1662db9)

**Reply by Navid200** [reply]
Can I see the logs please?  https://navid200.github.io/xDrip/docs/Logs.html

**Reply by BernhardRo** [reply]
do you get it, if I press "upload", or as screenshot?

**Reply by Navid200** [reply]
No, I will not get it if you upload.  Do you see any logs about the ui collector?

**Reply by BernhardRo** [reply]
yes, there are: UiBasedCollector
Receiving notifications that are not enabled to process: com.senseonics.gen12androidapp

**Reply by Navid200** [reply]
Have you allowed restricted settings for xDrip under Android settings?

**Reply by Navid200** [reply]
Wait, I don't get this!

Is that  log possibly an old log?
Can you please post a screenshot?

**Reply by jamorham** [reply]
It means the collector has been changed off UIBased I expect. We might need to widen the filter if this doesn't work or it might be presenting the data differently than expected.

**Reply by BernhardRo** [reply]
It should be base on the nightly:
![Screenshot_20240113-190147](https://github.com/NightscoutFoundation/xDrip/assets/35614076/1db10d72-4e5f-4e06-ad75-77eff0fea629)

**Reply by BernhardRo** [reply]
I switch between the UIBased collector and the one for Eversense

**Reply by Navid200** [reply]
> I switch between the UIBased collector and the one for Eversense  

Why do you switch?  That's confusing!
Do you get that log if you set the collector to Companion app?

**Reply by BernhardRo** [reply]
But that's just one entry. There are no further log entries regarding UIBased Collector

**Reply by BernhardRo** [reply]
Yes, for that log the collector was set to companion mode. I have it now since 1h in companion mode and do not get further log entries, but also no values

**Reply by Navid200** [reply]
OK, thanks

Please go back to your regular settings so that you can control your glucose.
In time, this will be addressed.  Either by me or by the lead developer JamOrHam.

**Reply by BernhardRo** [reply]
Can I support you further? I can imagine that it is very hard to debug something which is not local available.
By the way, the settings regarding notification access are looking like this
![Screenshot_20240113-200622](https://github.com/NightscoutFoundation/xDrip/assets/35614076/89944fb1-a4ae-405e-8757-726033f7df81)
It is strange that something is grayed out.

**Reply by Navid200** [reply]
Can you enable detail logs?  Although, I am not sure there are many.
Go to xDrip Settings, Less common settings, Extra Logging Settings, Extra tags for logging.
Enter the following.
`UiBasedCollector:v`  

Enable companion app.
Let it run for 5-10 minutes.
Do you see any new logs?

**Reply by jamorham** [reply]
@BernhardRo Please try the nightly B version just released some minutes ago. You may need to download it manually. See if that works?

**Reply by BernhardRo** [reply]
Yes, this works! Great job, thank you!
I need to find out if the US app has the same identifier. If so - fine! If not, than an additional identify is needed for it

**Reply by BernhardRo** [reply]
OK, I have checked it: the US version has the identifier "com.senseonics.androidapp". So this "gen12" of the EU version is missing. I do not have a running version available, but at least I know who to ask for testing!

**Comment by jamorham** [comment]
Yes this should be possible

**Comment by BernhardRo** [comment]
Have you already been able to extend the NotificationListenerService to the US version as well, so to notifications from "com.senseonics.androidapp"?

**Comment by 4RK4N** [comment]
been testing that as well. works good so far. sometimes it gets a bit out of sync like grabbing a value from the notification 30 seconds before a new one comes in so you then have an old value shown as new. but then it syncs itself quickly again by pushing out another value on notification change. i have now 120% of bg values grabbed per day which means it doing a bit too often but most of the time it works pretty well. good job.

good thing aaps has its smoothing plugins sorted out by now. without it it wouldnt be usable for looping.

question 1: since we have settings for the settings of everything can we have some settings to configure the xdrip smoothing algorithm and its weights? curently xdrips smoothing is way too weak to produce loopable values and needs another smoothing instance somewhere.

question 2: as you can see on BernhardRo's screenshot the transmitter battery value is also displayed in that eversense notification and aaps can display such a battery value now in the latest version. but the xdrip code doesnt look like its grabbing the battery value right now. can that be done?
